### PR TITLE
Enhance validation of dockstore.yml workflows

### PR DESCRIPTION
**Pull request by agduncan94**
https://github.com/dockstore/dockstore/issues/3410
Previously if a primary descriptor was not found for a dockstore yml workflow, the version is not added.

Now, the version is added and a validation message is associated with the dockstore.yml. If a test parameter file is missing, this is also kept.

Also adds a test for this situation.

Question for the group:
If a dockstore.yml is missing a test parameter file, it will currently cause the version to be invalid. Do we want this to be true?

### CODE REVIEW

**Comment by coverbeck** [comment]
> Question for the group:
> If a dockstore.yml is missing a test parameter file, it will currently cause the version to be invalid. Do we want this to be true?

Hmm, if a test parameter file is not required, then probably not. I see I made the key required, but did not require it to have any values. Probably makes sense to not require the key, although now the Java code will have to do null checks.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java [hunk 1]
```diff
@@ -704,20 +708,30 @@ private WorkflowVersion setupWorkflowFilesForGitHubVersion(Triple<String, Date,
...
                         if (!hasDuplicate) {
                             version.getSourceFiles().add(testJson);
                         }
+                    } else {
+                        missingParamFiles.add(testParameterPath);
                     }
                 }
+
+                if (missingParamFiles.size() > 0) {
+                    validationMessage.put("/.dockstore.yml", String.format("The following file(s) are missing: %s.",
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java:721]: Just my opinion:

I know it's a hassle to code and I know it's nit-picky, but I think we shouldn't use the `(s)` in messages -- we know if it's singular or plural.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java [hunk 2]
```diff
@@ -704,20 +708,30 @@ private WorkflowVersion setupWorkflowFilesForGitHubVersion(Triple<String, Date,
...
+                            missingParamFiles.stream().map(paramFile -> String.format("'%s'", paramFile)).collect(Collectors.joining(", "))));
+                }
             }
         } else {
             // File not found or null
             LOG.info("Could not find file " + primaryDescriptorPath + " in repo " + repository);
-            return null;
+            validationMessage.put("/.dockstore.yml", "Could not find the primary descriptor file '" + primaryDescriptorPath + "'.");
```
Comment (garyluu) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java:728]: Can you make "/.dockstore.yml" a constant somewhere?  There was a recent ticket saying how it sometimes was "/dockstore.yml" and it seems like now is as good a time as any to just start using a constant for that file path.

**Comment by Ldcabansay** [comment]
> @coverbeck So would a dockstore.yml not be invalid if one of the test parameter files listed in the dockstore.yml is not present, however we would display this as a warning message when the user is looking?
> 
> Edit: Maybe validations should change the boolean isValid to an enum. This enum could be VALID, INVALID, or WARNING. Thoughts?

On requiring test file: Seems like on manual register we don't require a test parameter file for registration. I'd vote for keeping this consistent. I think should complain if you do specify one and it can't find it though.

I like enum idea with warning as an option.

This is just validation for the .dockstore.yml file and not the workflow version itself, right? Where will this displayed on the info tab or versions tab? Bit concerned it could be confused with version validation.

**Comment by coverbeck** [comment]
> @coverbeck So would a dockstore.yml not be invalid if one of the test parameter files listed in the dockstore.yml is not present, however we would display this as a warning message when the user is looking?
> 
> Edit: Maybe validations should change the boolean isValid to an enum. This enum could be VALID, INVALID, or WARNING. Thoughts?

Yes, I think there should warnings without making it be invalid. Not sure if changing the column to an enum is the right way, but maybe. I don't know the whole data structure, but valid/invalid and warning off-hand seem to overlap or mean slightly different things. Could just be a wording thing though.

**Comment by agduncan94** [comment]
@Ldcabansay Those are good points. Currently, I have it so an invalid dockstore.yml => invalid version, which based on this discussion seems like that would only be if the primary descriptor is not found, and not if one or more test parameters are missing.

To try and summarise:
* dockstore.yml references descriptor that doesn't exist => invalid .dockstore.yml => invalid version
* dockstore.yml references a test file that doesn't exist => invalid .dockstore.yml => valid version

**Comment by agduncan94** [comment]
@denis-yuen any thoughts on the above comments?

**Comment by denis-yuen** [comment]
> @denis-yuen any thoughts on the above comments?

Seems reasonable, I think I was more concerned about how to debug white-space errors or more drastic typos, but this is good as well.

**Comment by agduncan94** [comment]
@denis-yuen those type of errors involve big changes with our error handling, which I don't think we will get to in this release, since we wouldn't be able to parse the file and wouldn't be able to associate it with a specific workflow.

**Comment by denis-yuen** [comment]
I would actually be ok with it reporting errors to the "wrong" base workflow without a workflow name. But I agree, it could be a pain

**Comment by agduncan94** [comment]
@denis-yuen rerunning to be sure.

**Comment by denis-yuen** [comment]
```
ailed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.4.1:enforce (enforce) on project dockstore: Execution enforce of goal org.apache.maven.plugins:maven-enforcer-plugin:1.4.1:enforce failed: Plugin org.apache.maven.plugins:maven-enforcer-plugin:1.4.1 or one of its dependencies could not be resolved: Failed to collect dependencies at org.apache.maven.plugins:maven-enforcer-plugin:jar:1.4.1 -> org.apache.maven.enforcer:enforcer-rules:jar:1.4.1 -> org.beanshell:bsh:jar:2.0b4: Failed to read artifact descriptor for org.beanshell:bsh:jar:2.0b4: Could not transfer artifact org.beanshell:bsh:pom:2.0b4 from/to central (https://repo.maven.apache.org/maven2): GET request of: org/beanshell/bsh/2.0b4/bsh-2.0b4.pom from central failed: Premature end of Content-Length delimited message body (expected: 1195; received: 0 
```
network issues? rre-running

**Pull request merged**